R/api.R

```diff
@@ -16,108 +16,327 @@
+  if (!is.null(max_items)) {
+    assert(is.numeric(max_items), "max_items must be a numeric type")
+    assert(max_items > 0,
+           "max_items must be greater than 0")
+    ids_list <- ids_list[1:min(max_items, length(ids_list))]
+  }
+  ids_list
+}
+
+#' @title Get Hacker News by Id
+#'
+#' @description Retrieves the item
+#' corresponding to specified id using Hacker News API
 #'
 #' @param id id of the item that should be retrieved
 #'
+#' @export
+#'
 #' @return item correspoing to the specified id
 #'
+#' @examples
+#'
+#' h_id <- get_max_item_id()
+#' hacker_article <- get_item_by_id(h_id)
+#'
 get_item_by_id <- function(id) {
+  assert(is.numeric(id) & id > 0 & id %% 1 == 0,
+         "id must be an integer greater than 0")
+
   request_url <- create_request_url(.base_url(), c("item", id))
-  .send_request(request_url)
+  item <- .send_request(request_url)
+  create_hn_item(item)
 }
 
-#' Get user corresponding to the specified username
+
+#' @title Get Hacker News user
 #'
-#' @export
+#' @description Get user corresponding to the specified username
 #'
 #' @param username of the user that should be retrieved
 #'
+#' @export
+#'
 #' @return user corresponding to the specified username
 #'
+#' @examples
+#' user <- get_user_by_username("steveddaniels")
+#' print(user)
 get_user_by_username <- function(username) {
+  assert(is.character(username), "username must be a character type object")
   request_url <- create_request_url(.base_url(), c("user", username))
-  .send_request(request_url)
+  user <- .send_request(request_url)
+  user <- create_hn_user(user)
+  user
 }
 
-#' Get the current largest item id
+#' @title Hacker News item largest id
+#'
+#' @description Retrives largest item id available via Hacker News API
 #'
 #' @export
 #'
 #' @return current largest item id
+#'
+#' @examples
+#' h_id <- get_max_item_id()
+#'
 get_max_item_id <- function() {
   request_url <- create_request_url(.base_url(), c("maxitem"))
   .send_request(request_url)
 }
 
-#' Get top stories
+#' @title Hacker News top stories ids
+#'
+#' @description Retrives top stories ids using Hacker News API
 #'
 #' @export
 #'
 #' @return list of top stories ids
-get_top_stories <- function() {
+#'
+#' @examples
+#' top_stories_ids <- get_top_stories_ids()
+#'
+get_top_stories_ids <- function() {
   request_url <- create_request_url(.base_url(), c("topstories"))
   .send_request(request_url)
 }
 
-#' Get new stories
+#' @title Hacker News top stories
+#'
+#' @description Retrives top stories using Hacker News API
+#'
+#' @param max_items Maximum number of items to retrive.
+#' If max_items = NULL, returns all available stories
+#'
+#' @export
+#'
+#' @return List of top stories
+#'
+#' @examples
+#' \dontrun{
+#' top_stories <- get_top_stories()
+#' top_5_stories <- get_top_stories(5)
+#' }
+get_top_stories <- function(max_items = NULL) {
+  ids_list <- get_top_stories_ids()
+  ids_list <- .trim_ids_list(ids_list, max_items)
+  lapply(ids_list, get_item_by_id)
```

Comment:
This code is often repeated. Why don't we create a function that accepts a vector of ids as input? :)

Reply:
so we will end up having 2 lines of code instead of 3?


# some assert on max_items
ids_list <- get_top_stories_ids()
 .get_items(ids_list, max_items)

Reply:
Yes, something like that. I would actually export this kind of function as it might be useful to some users. `get_items_by_ids(ids)`